File: Dynamic Programming/DP on string/Number of distinct subsequences/optimized_app.cpp
#include<bits/stdc++.h>
using namespace std;

// Time complextiy :- O(N)
// Space complexity :- O(N)

// Approach :-   
// for creating every subsequence possible subsequence we'll use take or not take case 

//Link :- https://www.geeksforgeeks.org/problems/number-of-distinct-subsequences0909/1

class Solution {
  public:
    int distinctSubseq(string &str) {
        int n = str.size(), mod = 1e9+7;
        vector<int> dp(n+1, 0), freq(26, -1);
        // base case 
        dp[0] = 1;
        
        for(int i=0; i<n; ++i){
            // 2 time of previous value due to take or not take value 
            dp[i+1] = (2*dp[i]) % mod;
            // if the char is repeating then substract duplicate ones 
            if(freq[str[i] - 'a'] != -1) 
                dp[i+1] = (dp[i+1] - dp[freq[str[i] - 'a']] + mod) % mod;
            
            // update the index of current char 
            freq[str[i] - 'a'] = i;
        }
        return dp[n];
    }
};

File: Queue/Minimum K Consecutive Bit Flips/optimized_app.java
import java.util.*

//Time complexity :- O(n)
//Space complexity :- O(n)

// Approach :- 
// we can keep track of number of time bit flip 
// if the current flip count is odd then the value should be 1 
// or if it's even then the current value should be 0
// we'll maintain a bool array to check where we flipped the bit  
// that help us to the track the validity of each flip till which index it's should bit flip
// i-k bit is flip then remove decrement the flip count 


//Link :- https://www.geeksforgeeks.org/problems/minimum-number-of-k-consecutive-bit-flips--171650/1

class Solution {
    public int kBitFlips(int[] arr, int k) {
        int n = arr.length;
        int flipped = 0, res = 0;
        boolean[] isFlipped = new boolean[n];
        
        for(int i=0; i<n; ++i){
            // validity of previous flip is expired 
            if(i-k >=0 && isFlipped[i-k] == true){
                flipped--;
            }
            // flip is needed 
            if(arr[i] == (flipped % 2)){
                // if index does not have sufficient next bit then it's not possible 
                if(i+k > n){
                    return -1;
                }
                // add the condition to flips 
                res += 1;
                flipped++;
                isFlipped[i] = true;
            }
        }
        return res;
    }
}
